### PR TITLE
discoverd: Reconnect event watches on error

### PR DESCRIPTION
**Pull request by lmars**
Related to #1607.

### CODE REVIEW

File: discoverd/client/watch.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+	"github.com/flynn/flynn/pkg/attempt"
+	"github.com/flynn/flynn/pkg/stream"
+)
+
+type WatchState string
+
+const (
+	WatchStateConnected    WatchState = "connected"
+	WatchStateDisconnected            = "disconnected"
```
Comment (titanous) [inline discoverd/client/watch.go]: This line needs a type annotation

File: discoverd/client/client.go [hunk 2]
```diff
@@ -20,7 +20,7 @@ type Service interface {
 	Instances() ([]*Instance, error)
 	Addrs() ([]string, error)
 	Leaders(chan *Instance) (stream.Stream, error)
-	Watch(events chan *Event) (stream.Stream, error)
+	Watch(events chan *Event) (*Watch, error)
```
Comment (titanous) [inline discoverd/client/client.go]: This should still return a `stream.Stream

File: discoverd/client/watch.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	return &Watch{
+		instances: make(map[string]*Instance),
+		done:      make(chan struct{}),
+	}
+}
+
+// SetStateChannel is used in tests to set a channel which will receive events
+// when the watch connects to / disconects from the server.
+func (w *Watch) SetStateChannel(ch chan WatchState) {
+	w.stateCh = ch
```
Comment (titanous) [inline discoverd/client/watch.go]: This isn't thread-safe. The comment should note it (making it only useful for tests) or it should use a `atomic.Value`.

File: discoverd/client/watch.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+var connectAttempts = attempt.Strategy{
+	Total: 30 * time.Second,
+	Delay: 200 * time.Millisecond,
+}
+
+// Watch starts an event stream and sends events to the given channel. It
+// reconnects to the server on error and generates any events which were missed
+// by comparing local state to the current state returned by the server.
```
Comment (titanous) [inline discoverd/client/watch.go]: Comment should note the retry logic.

File: discoverd/client/watch.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+					}
+
+					// the instance is known but has changed from when we last saw it, so send
+					// EventKindUpdate instead
+					event.Kind = EventKindUpdate
+				case EventKindUpdate:
+					watch.addInst(event.Instance)
+				case EventKindDown:
+					watch.delInst(event.Instance)
+				case EventKindServiceMeta:
```
Comment (titanous) [inline discoverd/client/watch.go]: Looks like leader events are missing?

**Comment by lmars** [comment]
@titanous comments addressed

File: discoverd/client/watch.go [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+					watch.delInst(event.Instance)
+				case EventKindLeader:
+					// don't send duplicate leader events
+					if !isCurrent && watch.leader != nil && watch.leader.ID == event.Instance.ID {
+						continue
+					}
+					watch.leader = event.Instance
+				case EventKindServiceMeta:
+					// don't send duplicate service meta events
+					if !isCurrent && reflect.DeepEqual(event.ServiceMeta, watch.serviceMeta) {
```
Comment (titanous) [inline discoverd/client/watch.go]: It'd be much more to just compare the Index.

**Comment by lmars** [comment]
@titanous comments addressed

**Comment by titanous** [comment]
LGTM

**Pull request merged**